Add "DynamicFunctionReturnTypeExtension" for "array_replace" + fix for "array_merge"

voku opened the pull request:
This Extension is a copy of "ArrayMergeFunctionDynamicReturnTypeExtension"

⇾ requested here: https://github.com/phpstan/phpstan-src/pull/622

INFO: Currently we are getting wrong types from "ArrayMergeFunctionDynamicReturnTypeExtension" because PHP handles array-keys different here: https://3v4l.org/YfKYq
- if the key is numeric, then the value will be added
- if the key is non-numeric, then the value will be replaced

ondrejmirtes commented on src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,67 @@
+				$argType = $argType->getIterableValueType();
+				if ($argType instanceof UnionType) {
+					foreach ($argType->getTypes() as $innerType) {
+						$argType = $innerType;
```
I don't understand this part - the last inner type in the union will win in the `$argType` variable?

ondrejmirtes replied:
I know the same code is in ArrayMergeFunctionDynamicReturnTypeExtension - it also puzzles me there.

ondrejmirtes replied:
:)

ondrejmirtes replied:
Let's try and see what happens :)

voku replied:
I re-added the fixes and added a limit for the array. I also added the fix for "array_merge" here, so we can see the differences between the implementations in one diff.

ondrejmirtes reviewed (changes requested):
How is this different to ArrayMergeFunctionDynamicReturnTypeExtension? Are they going to differ in the future?

voku committed:
Add "DynamicFunctionReturnTypeExtension" for "array_replace"

voku committed:
fix logic and "unpack" for "array_merge" and "array_replace"

ondrejmirtes reviewed (changes requested):
Please add a regression test for https://github.com/phpstan/phpstan/issues/5846 here and I'll take a look :) Thank you.

voku committed:
re-add logic for constant-arrays for "array_merge" and "array_replace"

+ try to prevent performance issues from it: maybe we need to adjust the consts "MAX_*"

voku commented:
> Please add a regression test for [phpstan/phpstan#5846](https://github.com/phpstan/phpstan/issues/5846) here and I'll take a look :) Thank you.

I re-added some changes, even more tests and added some checks to prevent performance issues, but maybe we need to adjust the consts "MAX_*" settings? I already tested the performance with a private project and blackfire.io and for me, it looks good, but as airways it depends on the tested code.
